File: 2021_12_24_advent_of_code_in_fsharp/AdventOfCode/AdventOfCode2021/Day01.fs
module AdventOfCode2021.Day01

open System

let loadData (input: string) = input.Split Environment.NewLine |> Array.map Int32.Parse

let countIncreases numbers = numbers |> Seq.pairwise |> Seq.filter (fun (prev, next) -> next > prev) |> Seq.length

let puzzle1 (input: string) =
    let numbers = loadData input
    let increases = countIncreases numbers
    string increases

let puzzle2 (input: string) =
    let numbers = loadData input
    let addedNumbers = numbers |> Seq.windowed 3 |> Seq.map Array.sum
    let increases = countIncreases addedNumbers
    increases |> string

File: 2022_04_23_advent_of_code_in_csharp/AdventOfCode/AdventOfCode2021/Day01.cs
namespace AdventOfCode.AdventOfCode2021;

static class Day1
{
    public static IEnumerable<int> LoadData(string input) =>
        input.Split(Environment.NewLine).Select(Int32.Parse);

    private static int CountIncreases(IEnumerable<int> numbers) =>
        numbers.Pairwise().Count(pair => pair.Item2 > pair.Item1);

    public static string Puzzle1(string input)
    {
        var numers = LoadData(input);
        var result = CountIncreases(numers);
        return result.ToString();
    }
    //=> input.Pipe(LoadData).Pipe(CountIncreases).ToString();

    public static string Puzzle2(string input)
    {
        var numers = LoadData(input);
        var addedNumbers = numers.Windowed(3).Select(Enumerable.Sum);
        var result = CountIncreases(addedNumbers);
        return result.ToString();
    }
}

